AI/VCAI/Goals/Explore.cpp

```diff
@@ -213,8 +196,7 @@ int Explore::howManyTilesWillBeDiscovered(
 			int3 npos = int3(x, y, pos.z);
 			if(cbp->isInTheMap(npos)
 				&& pos.dist2d(npos) - 0.5 < radious
-				&& !ts->fogOfWarMap[npos.x][npos.y][npos.z]
-				&& hasReachableNeighbor(npos, h, cbp, aip))
```

Comment:
Why do you think it is wrong? It is dead end exploring cancellation.

Reply:
This function counts unexplored tiles within sight radius. `hasReachableNeighbor` check does not make sense.

Reply:
It checks if unexplored tile has reachable neighbor an thus can be visited later. Otherwise there is much lower probability it will be useful. Even if it is not just a rock but a tile behind some border. The only way to reach it in such case is to cast air walking or dimension door. AI can not do both for now.

Reply:
It might be an option to count potentially visitable tile as 1 and potentially unreachable as 0.1 or something. If it behaves badly somewhere lets look onto particular case.

Reply:
If tile is not visible yet it is not reachable too. Reachable neighbour check only works for border of FoW.

Reply:
Yes, it is an expected behavior. The idea is that AI should not get close to border because this way it explores behind it and thus explores unreachable area.

Reply:
Does it hurt performance? We can try to check if there is any visitable tile but do not require it being reachable. This way we will cancel exploration of rocks. But it requires cb->getTile()

Reply:
Basically all changes you see in this branch are about what Alex found in Valgrind's Callgrind output.

Reply:
Just removing code is not always the best action to optimize things. I am afraid AI might start spending more turns on exploring thus being less productive. I have some ideas to optimize Explore logically by introducing the middle stage when we scan an area around hero instead of all map. Having such stage allows to disable this check in full scan. Also Explore can try to acquire kind of tile accessibility matrix which will be allocated and cached in PathfindingManager

Reply:
Comment out this function. I will try to find a better solution. We just need to reduce scan area somehow. And I have a few ideas.

Reply:
I can not remove `howManyTilesWillBeDiscovered` entirely. It is used in both `explorationScanRange` and `explorationBestNeighbour`

Reply:
I mean hasReachableNeighbor

Reply:
done

Reply:
Thanks! I will try to do something about it in the nearest future.

Reply:
Wait, this piece of code replaced old one, improving both AI performance and speed. I don't quite see how you could remove it entirely.

Reply:
hasReachableNeighbor is for dead end cancellation. It improves exploration behavior but obviously not performance

Reply:
What if there is a quest artefact hidden in the mountains (only accessible by the Door by design), will AI be able to find it with dead end cancellation?

Reply:
Good question. Still I feel like dead end cancellation is needed but at the same time there should be a way to explore all possible tiles as low priority task.

Reply:
Lets add dead end check only to `explorationBestNeighbour`. So hero will efficiently continue exploring but start at any tile.

Reply:
This does not work 100% nice. Sure it helps at certain extent. AI goes to dead end, cancels it and now he has nothing to explore because in front is dead end and behind is already explored. It enables full scan without dead-end cancellation and finds tiles in front so he goes and explores them because they are very close.

Reply:
It is very tricky, this exploration. Maybe with middle stage where we scan in range around hero with dead end cancellation and only after we we enable to find a target near we run full scan with disabled dead end we can achieve better behavior.

Reply:
Also as an option we can cancel exploration for a whole turn if it once yield no results. Also exploration sometimes triggered twice or even more doing exactly the same thing. Conquer, GatherArmy and CollectRes can produce exploration for instance. Explore decomposes to Explore(hero) instead of directly returning found goals. Also as an option we can put Explore to the list of basic goals and remove it from everywhere else thus allowing to decompose it only once per hero move. Also we can move hero not one tile at a time but 3 tiles or half of exploration radius thus reducing amount of decomposition cycles

Reply:
Also might be an option to check if vcmieagles ai is set and disable explore at all

Reply:
Can we just skip this line for this PR and test non-controversial performance optimizations?

Reply:
Comment it out, already done I suppose and I will think about Explore optimization later. I expect I have to establish some profiler locally on Windows and will be able to see what is going on myself.

Reply:
There's already heroesUnableToExplore container with logic exactly for that.
However, it's difficult to exclude exploration at all. If AI wants to explore, it means either it didn't explore entire map yet or map contains no necessary objectives at all.

Reply:
Probably you are right, it should not take much time if there is no visible tiles. The worst case probably is 50-75% open map

Reply:
What is "commented out", the function call is still removed in this PR.

Reply:
I have a few ideas for myself.
1. add a middle stage where AI scans 20x20 area around hero for a new point
2. scan perimeter of visible area before expanding it by sight radius and if it yields any result - return it. If all perimeter tiles are blocked expand and scan with deadend cancellation disabled.
3. might be worth to split AI pathfinder API onto one updatePaths(std:vector<HeroPtr>) and the rest is const. This way we can drop synchronization using mutex at all I think.

Reply:
Reverted. Any further improvements considered out of scope.